src/motion_input_shaping/zero_vibration_stream_generator.py

```diff
@@ -0,0 +1,361 @@
+        return self._damping_ratio
+
+    @damping_ratio.setter
+    def damping_ratio(self, value: float) -> None:
+        """Set the target damping ratio for input shaping."""
+        if value < 0:
+            raise ValueError(
+                f"Invalid damping ratio: {value}. Value must be greater than or equal to 0."
+            )
+        self._damping_ratio = value
+        self._impulses_updated = False
+
+    @property
+    def shaper_type(self) -> ShaperType:
+        """Get input shaper type."""
+        return self._shaper_type
+
+    @shaper_type.setter
+    def shaper_type(self, value: ShaperType) -> None:
+        """Set input shaper type."""
+        self._shaper_type = value
+        self._impulses_updated = False
+
+    def get_impulses(self) -> list[float]:
+        """Get shaper impulse magnitudes."""
+        if self._impulses_updated is False:
+            self._update_shaper_impulses()
+        return self._impulses
+
+    def get_impulse_times(self) -> list[float]:
+        """Get shaper impulse times."""
+        if self._impulses_updated is False:
+            self._update_shaper_impulses()
+        return self._impulse_times
+
+    def _update_shaper_impulses(self) -> None:
+        """
+        Calculate times and unitless magnitude of impulses to perform the input shaping.
+
+        The sum of all impulses should total to 1 to maintain the same final state.
+        """
+        k = math.exp(
+            (-1 * math.pi * self.damping_ratio) / math.sqrt(1 - self.damping_ratio**2)
+        )  # Decay factor
+
+        match self.shaper_type:
+            case ShaperType.ZV:
+                a0 = 1 / (1 + k)
+                a1 = k / (1 + k)
+                t0 = 0
+                t1 = self.resonant_period / 2
+                self._impulses = [a0, a1]
+                self._impulse_times = [t0, t1]
+            case ShaperType.ZVD:
+                a0 = 1 / (1 + 2 * k + k**2)
+                a1 = 2 * k / (1 + 2 * k + k**2)
+                a2 = k**2 / (1 + 2 * k + k**2)
+                t0 = 0
+                t1 = self.resonant_period / 2
+                t2 = self.resonant_period
+                self._impulses = [a0, a1, a2]
+                self._impulse_times = [t0, t1, t2]
+            case ShaperType.ZVDD:
+                a0 = 1 / (1 + 3 * k + 3 * k**2 + k**3)
+                a1 = 3 * k / (1 + 3 * k + 3 * k**2 + k**3)
+                a2 = 3 * k**2 / (1 + 3 * k + 3 * k**2 + k**3)
+                a3 = k**3 / (1 + 3 * k + 3 * k**2 + k**3)
+                t0 = 0
+                t1 = self.resonant_period / 2
+                t2 = self.resonant_period
+                t3 = self.resonant_period * 3 / 2
+                self._impulses = [a0, a1, a2, a3]
+                self._impulse_times = [t0, t1, t2, t3]
+            case _:
+                raise ValueError(f"Shaper type {self.shaper_type} is not valid.")
+        self._impulses_updated = True
+
+    def shape_trapezoidal_motion(
```

Comment:
A ton of these function definitions are extremely close to zero_vibration_shaper.py. I feel like there's so much redundant code. I wonder if n and calculate_n could be private in the other class? There's a chance we might be able to create fully identical class definitions for both classes, and maybe even simplify them to one class? 

Maybe it's not worth it. I just hate duplicating so much identical code.

Reply:
There are a lot of elements shared between them but what is being done is also very different. For example, the impulse calculations look very similar but none of it can be shared between the two. I thought that keeping them separate is less confusing in a code example as it can become difficult to figure out which bits are actually relevant for a particular mode. I also see the argument for combining it all though.

One option to reduce the identical code could be to make a parent class for the stuff that is common between the two ( `resonant_frequency`, `damping_ratio`, and `resonant_period`).

Reply:
`resonant_frequency`, `damping_ratio`, and `resonant_period` has been moved to a Plant class. Impulse calculations are still kept separate as they are not cross compatible.